### DISCUSSION TITLE
data race in `shell.c`

### DISCUSSION THREAD

**Discussion by ghost** [zephyrproject-rtos/zephyr#64632] [General]
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/shell/shell.c#L1613

If, I am not mistaken. Line 1613 in `shell.c` can lead to a datarace. Expectation to caller is unclear. For example, code like this is a datarace?:

```C
static char prompt[256];
strcpy(prompt, "Hello");
shell_prompt_change(sh, prompt);
```

because it boils down to:

```C
char *a;
void shell_thread(void) {
    while(1) {
        // reads a
        if (a != NULL) {
            printf("%s\n", a);
        }
    }
}

void caller_thread(void) {
    static char prompt[256];
    strcpy(prompt, "Hello");
    a = prompt; // equivalent of shell_prompt_change(sh, prompt);
}
```

Could someone confirm, that this is the case? If so, I can create an issue.
Thanks in advance.

**Comment by nordicjm** [comment]
@jakub-uC

**Comment by ghost** [comment]
Basically:
caller_thread could be concurrently writing to the prompt (optimizer is alowed to swap `a=prompt;` and `strncpy`)
shell_thread could be concurrently reading it.
It is not possible to provide a synchronization mechanism.

Proposed solution would be, that `shell_prompt_change` writes the data to private buffer and cares about the synchronisation with the shell_thread.

**Comment by ghost** [comment]
@nordicjm Any news on this? May I create the issue?

**Reply by nordicjm** [reply]
Feel free to make the issue, Jakub is the shell maintainer so was bringing this to their attention, they'd need to review the bug

**Comment by ghost** [comment]
https://github.com/zephyrproject-rtos/zephyr/issues/64972 created issue from this
